rust/benchmark-datasets/src/types.rs

```diff
@@ -0,0 +1,235 @@
+            let file_name = format!(
+                "frozen_query_subset_{}_{}_{}.bin",
+                corpus_dataset.get_name(),
+                min_results_per_query,
+                max_num_of_queries
+            );
+
+            let file = get_or_populate_cached_dataset_file(
+                self.get_name(),
+                &file_name,
+                Some(cache_dir.unwrap_or(get_dir_for_persistent_dataset_files())),
+                |mut file| async move {
+                    let (corpus_index, corpus_index_schema) = futures::StreamExt::boxed_local(
+                        corpus_dataset.create_documents_stream().await?.filter_map(
+                            |doc| match doc {
+                                Ok(doc) => Some(doc),
+                                Err(_) => None,
+                            },
+                        ),
+                    )
+                    .build_tantivy_index()
+                    .await?;
+
+                    let mut shuffled_queries = self
+                        .create_documents_stream()
+                        .await?
+                        .filter_map(|doc| match doc {
+                            Ok(doc) => Some(doc.content),
+                            Err(_) => None,
+                        })
+                        .collect::<Vec<String>>()
+                        .await;
+                    shuffled_queries.shuffle(&mut rand::thread_rng());
+
+                    let reader = corpus_index
+                        .reader_builder()
+                        .reload_policy(ReloadPolicy::OnCommit)
+                        .try_into()?;
+
+                    let searcher = reader.searcher();
+                    let query_parser = QueryParser::for_index(
+                        &corpus_index,
+                        vec![corpus_index_schema.get_field("content")?],
+                    );
+
+                    let mut frozen_query_subset = FrozenQuerySubset {
+                        queries: Vec::new(),
+                        queries_from_dataset: self.get_name().to_string(),
+                        dataset_queries_tested_against: corpus_dataset.get_name().to_string(),
+                        min_results_per_query,
+                    };
+
+                    for query_text in shuffled_queries.iter() {
+                        let query = query_parser.parse_query(query_text)?;
+                        let top_docs =
+                            searcher.search(&query, &TopDocs::with_limit(min_results_per_query))?;
+                        if top_docs.len() >= min_results_per_query {
+                            frozen_query_subset.queries.push(query_text.to_string());
+                        }
+
+                        if frozen_query_subset.queries.len() == max_num_of_queries {
+                            break;
+                        }
+                    }
+
+                    let serialized = bincode::serialize(&frozen_query_subset)?;
```

Comment:
q: why bincode?

Reply:
I just wanted a simple way to serialize/de-serialize a struct. Maybe there's a better option?

Reply:
fair enough wondering if we considered other things but not critical here